addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,434 @@
+}
+
+// PkgiStatusChanged returns a predicate.Predicate that filters pkgi objects which their status has changed
+func (r *PackageInstallStatusReconciler) PkgiStatusChanged(log logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "createEvent"))
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return r.processPkgiStatus(e.ObjectNew, log.WithValues("predicate", "updateEvent"))
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "deleteEvent"))
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "genericEvent"))
+		},
+	}
+}
+
+// processPkgiStatus returns true if pkgi status should be processed.
+// pkgi status can be processed if it is a managed package and exists in the list of packages defined in ClusterBootstrap
+func (r *PackageInstallStatusReconciler) processPkgiStatus(o client.Object, log logr.Logger) bool {
+	var pkgi *kapppkgiv1alpha1.PackageInstall
+	switch obj := o.(type) {
+	case *kapppkgiv1alpha1.PackageInstall:
+		pkgi = obj
+	default:
+		//TODO: check this
+		// Defaults to true so we don't filter out other objects as the filters are global
+		log.Info("Expected object type of PackageInstall. Got object type", "actualType", fmt.Sprintf("%T", o))
+		return true
+	}
+
+	clusterNamespace, clusterName := r.getClusterNamespacedName(pkgi)
+	if clusterNamespace == "" || clusterName == "" {
+		return false
+	}
+
+	isManaged, err := r.isPackageManaged(clusterName, pkgi.Name, clusterNamespace)
+	if err != nil {
+		return false
+	}
+
+	return isManaged
+}
+
+// isPackageManaged checks if the provided PackageInstall is among the list of managed(core/additional) packages
+func (r *PackageInstallStatusReconciler) isPackageManaged(clusterName, pkgiName, pkgiNamespace string) (bool, error) {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	clusterKey := client.ObjectKey{Namespace: pkgiNamespace, Name: clusterName}
+	if err := r.Client.Get(r.Ctx, clusterKey, clusterBootstrap); err != nil {
+		r.Log.Error(err, fmt.Sprintf("error getting ClusterBootstrap resource for cluster '%s/%s'", clusterKey.Namespace, clusterKey.Name))
+		return false, err
+	}
+
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+		clusterBootstrap.Spec.Kapp,
+	}, clusterBootstrap.Spec.CNIs...)
+	packages = append(packages, clusterBootstrap.Spec.AdditionalPackages...)
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		// ensure the name of the PackageInstall matches the name of the managed packages in the CLusterBootstrap resource
+		if pkgiName == util.GeneratePackageInstallName(clusterName, pkg.RefName) {
```

Comment:
question:  I started thinking if there is any way we could know if that PKGI is managed or not in O(1) time complexity. Would a label/annotation help here ? If no label/annotation is available, then we fall back to this O(n) solution ? Current code definitely works, but just have some concern on the performance.

Reply:
I double checked if we can add a label for this and seems like we can't do that

Reply:
Is the `ClusterNameLabel` and `ClusterNamespaceLabel` sufficient for this purpose?

Reply:
I mean we have added these 2 labels. Are they enough to indicate the pkgi is managed? I assume others wouldn't add such labels.

Reply:
true, but @vijaykatam wanted to use existence of the package in the ClusterBootstrap as the source of truth, and don't rely on labels/annotations